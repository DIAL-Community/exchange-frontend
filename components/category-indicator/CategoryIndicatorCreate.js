import { useCallback } from 'react'
import { useIntl } from 'react-intl'
import { useQuery } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'
import Breadcrumb from '../shared/Breadcrumb'
import { handleMissingData, handleQueryError } from '../shared/GraphQueryHandler'
import { RUBRIC_CATEGORY_QUERY } from '../shared/query/categoryIndicator'
import CategoryIndicatorForm from './fragments/CategoryIndicatorForm'
import CategoryIndicatorSimpleLeft from './fragments/CategoryIndicatorSimpleLeft'

const CategoryIndicatorCreate = ({ categorySlug }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const { loading, error, data } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { categorySlug },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.CREATING
      }
    }
  })

  if (loading) {
    return handleMissingData()
  } else if (error) {
    return handleQueryError(error)
  } else if (!data?.rubricCategory) {
    return handleMissingData()
  }

  const { rubricCategory } = data

  const slugNameMapping = (() => {
    const map = {
      create: format('app.create')
    }

    map[rubricCategory.slug] = rubricCategory.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping} />
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3'>
          <CategoryIndicatorSimpleLeft rubricCategory={rubricCategory} />
        </div>
        <div className='lg:basis-2/3'>
          <CategoryIndicatorForm rubricCategory={rubricCategory} />
        </div>
      </div>
    </div>
  )
}

export default CategoryIndicatorCreate
