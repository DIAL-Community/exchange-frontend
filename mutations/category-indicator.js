import { gql } from '@apollo/client'

export const CREATE_CATEGORY_INDICATOR = gql`
  mutation CreateCategoryIndicator (
    $name: String!
    $slug: String!
    $rubricCategorySlug: String
    $indicatorType: String!
    $dataSource: String
    $scriptName: String
    $weight: Float!
    $description: String!
  ) {
    createCategoryIndicator(
      name: $name
      slug: $slug
      rubricCategorySlug: $rubricCategorySlug
      indicatorType: $indicatorType
      dataSource: $dataSource
      scriptName: $scriptName
      weight: $weight
      description: $description
    ) {
      categoryIndicator{
        id
        slug
      }
      errors
    }
  }
`

export const DELETE_CATEGORY_INDICATOR = gql`
  mutation DeleteCategoryIndicator($id: ID!) {
    deleteCategoryIndicator(id: $id) {
      rubricCategorySlug
    }
  }
`
