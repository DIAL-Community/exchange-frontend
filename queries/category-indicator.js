import { gql } from '@apollo/client'

export const CATEGORY_INDICATOR_QUERY = gql`
  query CategoryIndicator($slug: String!) {
    categoryIndicator(slug: $slug) {
      slug
      name
      id
      indicatorType
      weight
      dataSource
      scriptName
      rubricCategoryId
      categoryIndicatorDescription {
        description
      } 
    }
  }
`
