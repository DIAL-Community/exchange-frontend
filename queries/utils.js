const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
