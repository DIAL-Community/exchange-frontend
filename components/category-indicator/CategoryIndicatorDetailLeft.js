import Bookmark from '../shared/common/Bookmark'
import Comment from '../shared/common/Comment'
import Share from '../shared/common/Share'
import { ObjectType } from '../utils/constants'
import CategoryIndicatorDetailHeader from './fragments/CategoryIndicatorDetailHeader'
import CategoryIndicatorDetailNav from './fragments/CategoryIndicatorDetailNav'

const CategoryIndicatorDetailLeft = ({ scrollRef, rubricCategory, categoryIndicator }) => {
  return (
    <div className='bg-dial-slate-100 lg:h-full'>
      <div className='flex flex-col gap-y-3 px-4 lg:px-6 lg:py-3'>
        <CategoryIndicatorDetailHeader
          rubricCategory={rubricCategory}
          categoryIndicator={categoryIndicator}
        />
        <hr className='border-b border-dial-slate-200'/>
        <CategoryIndicatorDetailNav
          rubricCategory={rubricCategory}
          categoryIndicator={categoryIndicator}
          scrollRef={scrollRef}
        />
        <hr className='border-b border-dial-slate-200'/>
        <div className='hidden lg:flex flex-col gap-y-3'>
          <Bookmark object={categoryIndicator} objectType={ObjectType.CATEGORY_INDICATOR} />
          <hr className='border-b border-dial-slate-200'/>
          <Share />
          <hr className='border-b border-dial-slate-200'/>
          <Comment entityKey={'categoryIndicator.label'} scrollRef={scrollRef} />
          <hr className='border-b border-dial-slate-200'/>
        </div>
      </div>
    </div>
  )
}

export default CategoryIndicatorDetailLeft
