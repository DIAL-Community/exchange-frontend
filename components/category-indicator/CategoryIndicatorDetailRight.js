import { forwardRef, useCallback, useImperativeHandle, useRef } from 'react'
import { useIntl } from 'react-intl'
import CommentsSection from '../shared/comment/CommentsSection'
import Bookmark from '../shared/common/Bookmark'
import Share from '../shared/common/Share'
import EditButton from '../shared/form/EditButton'
import { HtmlViewer } from '../shared/form/HtmlViewer'
import { ObjectType } from '../utils/constants'
import DeleteCategoryIndicator from './fragments/DeleteCategoryIndicator'

const CategoryIndicatorDetailRight = forwardRef(({ categoryIndicator, editingAllowed, deletingAllowed }, ref) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const descRef = useRef()
  const categoryIndicatorRef = useRef()
  const commentsSectionRef = useRef()
  useImperativeHandle(
    ref,
    () => [
      { value: 'ui.common.detail.description', ref: descRef },
      { value: 'categoryIndicator.header', ref: categoryIndicatorRef },
      { value: 'ui.comment.label', ref: commentsSectionRef }
    ],
    []
  )

  const editPath = `${categoryIndicator.slug}/edit`

  return (
    <div className='px-4 lg:px-0 py-4 lg:py-6'>
      <div className='flex flex-col gap-y-3'>
        <div className='flex gap-x-3 ml-auto'>
          { editingAllowed && <EditButton type='link' href={editPath} /> }
          { deletingAllowed && <DeleteCategoryIndicator categoryIndicator={categoryIndicator} /> }
        </div>
        <div className='text-xl font-semibold text-dial-plum py-3' ref={descRef}>
          {format('ui.common.detail.description')}
        </div>
        <div className='block'>
          <HtmlViewer
            initialContent={categoryIndicator?.categoryIndicatorDescription?.description}
            editorId='rubric-category-description'
          />
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='flex flex-col gap-y-3 text-sm'>
          <div className='text-xl font-semibold text-dial-plum'>
            {format('categoryIndicator.dataSource')}
          </div>
          {categoryIndicator?.dataSource}
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='flex flex-col gap-y-3 text-sm'>
          <div className='text-xl font-semibold text-dial-plum'>
            {format('categoryIndicator.weight')}
          </div>
          {categoryIndicator?.weight}
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='flex flex-col gap-y-3 text-sm'>
          <div className='text-xl font-semibold text-dial-plum'>
            {format('categoryIndicator.indicatorType')}
          </div>
          {categoryIndicator?.indicatorType}
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='flex flex-col gap-y-3 text-sm'>
          <div className='text-xl font-semibold text-dial-plum'>
            {format('categoryIndicator.scriptName')}
          </div>
          {categoryIndicator?.scriptName ?? format('general.na')}
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='lg:hidden flex flex-col gap-y-3'>
          <Bookmark object={categoryIndicator} objectType={ObjectType.CATEGORY_INDICATOR} />
          <hr className='border-b border-dial-slate-200'/>
          <Share />
          <hr className='border-b border-dial-slate-200'/>
        </div>
        <CommentsSection
          commentsSectionRef={commentsSectionRef}
          objectId={categoryIndicator.id}
          objectType={ObjectType.CATEGORY_INDICATOR}
        />
      </div>
    </div>
  )
})

CategoryIndicatorDetailRight.displayName = 'CategoryIndicatorDetailRight'

export default CategoryIndicatorDetailRight
