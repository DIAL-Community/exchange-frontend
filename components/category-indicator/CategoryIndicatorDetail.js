import { useRef, useState } from 'react'
import { useApolloClient, useQuery } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT, GRAPH_QUERY_POLICY_SLUG } from '../../lib/apolloClient'
import Breadcrumb from '../shared/Breadcrumb'
import { handleLoadingQuery, handleMissingData, handleQueryError } from '../shared/GraphQueryHandler'
import { CATEGORY_INDICATOR_POLICY_QUERY, CATEGORY_INDICATOR_QUERY } from '../shared/query/categoryIndicator'
import { fetchOperationPolicies } from '../utils/policy'
import CategoryIndicatorDetailLeft from './CategoryIndicatorDetailLeft'
import CategoryIndicatorDetailRight from './CategoryIndicatorDetailRight'

const CategoryIndicatorDetail = ({ categorySlug, indicatorSlug }) => {
  const scrollRef = useRef(null)
  const client = useApolloClient()

  const [editingAllowed, setEditingAllowed] = useState(false)
  const [deletingAllowed, setDeletingAllowed] = useState(false)

  const { loading, error, data } = useQuery(CATEGORY_INDICATOR_QUERY, {
    variables: { categorySlug, indicatorSlug },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  if (loading) {
    return handleLoadingQuery()
  } else if (error) {
    return handleQueryError(error)
  } else if (!data?.categoryIndicator && !data.rubricCategory) {
    return handleMissingData()
  }

  const { categoryIndicator, rubricCategory } = data

  fetchOperationPolicies(
    client,
    CATEGORY_INDICATOR_POLICY_QUERY,
    ['editing', 'deleting'],
    { categorySlug: GRAPH_QUERY_POLICY_SLUG, indicatorSlug: GRAPH_QUERY_POLICY_SLUG }
  ).then(policies => {
    setEditingAllowed(policies['editing'])
    setDeletingAllowed(policies['deleting'])
  })

  const slugNameMapping = (() => {
    const map = {}
    map[rubricCategory.slug] = rubricCategory.name
    map[categoryIndicator.slug] = categoryIndicator.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping} />
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3 shrink-0'>
          <CategoryIndicatorDetailLeft
            scrollRef={scrollRef}
            rubricCategory={rubricCategory}
            categoryIndicator={categoryIndicator}
          />
        </div>
        <div className='lg:basis-2/3 shrink-0'>
          <CategoryIndicatorDetailRight
            ref={scrollRef}
            rubricCategory={rubricCategory}
            categoryIndicator={categoryIndicator}
            editingAllowed={editingAllowed}
            deletingAllowed={deletingAllowed}
          />
        </div>
      </div>
    </div>
  )
}

export default CategoryIndicatorDetail
