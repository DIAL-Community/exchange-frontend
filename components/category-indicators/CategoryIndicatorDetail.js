import { useQuery } from '@apollo/client'
import { useEffect, useMemo } from 'react'
import NotFound from '../shared/NotFound'
import { Error, Loading } from '../shared/FetchStatus'
import { CATEGORY_INDICATOR_QUERY } from '../../queries/category-indicator'
import { RUBRIC_CATEGORY_QUERY } from '../../queries/rubric-category'
import CategoryIndicatorDetailLeft from './CategoryIndicatorDetailLeft'
import CategoryIndicatorDetailRight from './CategoryIndicatorDetailRight'

const CategoryIndicatorDetail = ({ rubricCategorySlug, categoryIndicatorSlug, locale }) => {
  const { loading, error, data: categoryIndicatorData, refetch } = useQuery(CATEGORY_INDICATOR_QUERY, {
    variables: { slug: categoryIndicatorSlug },
    context: { headers: { 'Accept-Language': locale } }
  })

  const { data: rubricCategoryData } = useQuery(RUBRIC_CATEGORY_QUERY, { variables: { slug: rubricCategorySlug } })

  const slugNameMapping = useMemo(
    () => ({
      [categoryIndicatorData?.categoryIndicator.slug]: categoryIndicatorData?.categoryIndicator.name,
      [rubricCategoryData?.rubricCategory.slug]: rubricCategoryData?.rubricCategory.name
    }),
    [categoryIndicatorData?.categoryIndicator, rubricCategoryData?.rubricCategory]
  )

  useEffect(() => {
    refetch()
  }, [refetch, locale])

  if (loading) {
    return <Loading />
  } else if (error && error.networkError) {
    return <Error />
  } else if (error && !error.networkError) {
    return <NotFound />
  }

  return (
    <div className='flex flex-col lg:flex-row pb-8'>
      <div className='lg:sticky lg:top-66px w-full lg:w-1/3 xl:w-1/4 h-full p-4'>
        <CategoryIndicatorDetailLeft categoryIndicator={categoryIndicatorData?.categoryIndicator} slugNameMapping={slugNameMapping} />
      </div>
      <div className='w-full lg:w-2/3 xl:w-3/4'>
        <CategoryIndicatorDetailRight categoryIndicator={categoryIndicatorData?.categoryIndicator} slugNameMapping={slugNameMapping} />
      </div>
    </div>
  )
}

export default CategoryIndicatorDetail
