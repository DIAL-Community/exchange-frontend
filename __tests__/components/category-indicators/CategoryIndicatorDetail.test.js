import { render, waitForAllEffects } from '../../test-utils'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import CategoryIndicatorDetail from '../../../components/category-indicators/CategoryIndicatorDetail'
import { CATEGORY_INDICATOR_QUERY } from '../../../queries/category-indicator'
import { mockNextUseRouter } from '../../utils/nextMockImplementation'
import { categoryIndicator } from './data/CategoryIndicatorDetail'

mockNextUseRouter()
describe('Unit test for the CategoryIndicatorDetail component.', () => {
  const mockCategoryIndicator = generateMockApolloData(CATEGORY_INDICATOR_QUERY, { slug: 'test_category_indicator' }, null, categoryIndicator)

  test('Should render CategoryIndicatorDetail component', async () => {
    const { container } = render(
      <CustomMockedProvider mocks={[mockCategoryIndicator]}>
        <CategoryIndicatorDetail slug='test_category_indicator'/>
      </CustomMockedProvider>
    )
    await waitForAllEffects()
    expect(container).toMatchSnapshot()
  })
})
