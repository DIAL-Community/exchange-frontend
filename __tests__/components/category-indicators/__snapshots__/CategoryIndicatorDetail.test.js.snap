// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the CategoryIndicatorDetail component. Should render CategoryIndicatorDetail component 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col lg:flex-row pb-8 max-w-catalog mx-auto"
  >
    <div
      class="lg:sticky lg:top-66px w-full lg:w-1/3 xl:w-1/4 h-full p-4"
    >
      <div
        class="block lg:hidden"
      >
        <div
          class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
        >
          <a
            class="inline text-dial-blue h5"
            href="/"
          >
            Home
          </a>
          <div
            class="inline h5"
          >
             &gt; 
            <a
              class="text-dial-gray-dark"
              href="/"
            />
          </div>
        </div>
      </div>
      <div
        class="h-12 w-full inline-flex gap-3 items-center"
      />
      <div
        class="bg-white border-2 border-dial-gray lg:mr-6 shadow-lg"
      >
        <div
          class="flex flex-col p-4 text-dial-gray-dark"
        >
          <div
            class="font-semibold text-lg"
          >
            Test Category Indicator
          </div>
          <hr
            class="my-2"
          />
          <div
            class="text-sm"
          >
            <div>
              <span
                class="font-semibold"
              >
                Weight
              </span>
              : 
              0.9
            </div>
            <div>
              <span
                class="font-semibold"
              >
                Indicator type
              </span>
              : 
              numeric
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full lg:w-2/3 xl:w-3/4"
    >
      <div
        class="px-4 text-dial-gray-dark"
      >
        <div
          class="hidden lg:block"
        >
          <div
            class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
          >
            <a
              class="inline text-dial-blue h5"
              href="/"
            >
              Home
            </a>
            <div
              class="inline h5"
            >
               &gt; 
              <a
                class="text-dial-gray-dark"
                href="/"
              />
            </div>
          </div>
        </div>
        <div
          class="mb-8"
        >
          <div
            class="card-title mb-3"
          >
            Data source
          </div>
          Test data source
        </div>
        <div
          class="mb-8"
        >
          <div
            class="card-title mb-3"
          >
            Script name
          </div>
          Test script name
        </div>
        <div>
          <div
            class="card-title mb-3"
          >
            Description
          </div>
          Test description
        </div>
      </div>
    </div>
  </div>
</div>
`;
